File: hub/src/light/effect/color_effects/triple_cubic_color_transition.py
from light.effect.partial_effect import PartialEffect
from light.effect.resolved_effect import ResolvedEffect
from util.color_transition import ColorTransition

class TripleCubicColorTransition(PartialEffect):
    def __init__(self, srcRgb, dstRgb1, dstRgb2, duration):
        super().__init__(0)
        self.src = srcRgb
        self.dst1 = dstRgb1
        self.dst2 = dstRgb2
        self.duration = duration
    
    def getEffect(self, t):
        dt = t - self.startTime
        tNorm = 1 - dt / self.duration
        tNorm = (tNorm * tNorm * tNorm)

        dst = self.dst1 if tNorm > 0.5 else self.dst2
        src = self.src if tNorm > 0.5 else self.dst1
        tNormAdjusted = (tNorm * 2) if tNorm <= 0.5 else ((tNorm - 0.5) * 2)
        
        return ResolvedEffect.createRgbw(ColorTransition.linear(1 - tNormAdjusted, src, dst))

    def isTemporal(self):
        return False

    def isComplete(self, t):
        return False

File: hub/drumlite-js/src/light/effect/color/TripleCubicColorTransition.ts
import RGB from "../../RGB"
import PartialEffect, { EffectParameters, EffectParameter } from "../PartialEffect"
import ResolvedEffect from "../../../effect/ResolvedEffect";
import ColorTransition from "../../../util/ColorTransition";

export class TripleCubicColorTransitionParams extends EffectParameters {
    effectName = "Triple Cubic Transition";
    typeName = "Color";

    constructor(src: RGB = new RGB(), dst1: RGB = new RGB(), dst2: RGB = new RGB(), duration: number = 1) {
        super(0);
        this.params.src = new EffectParameter<RGB>("Start Color", src);
        this.params.dst1 = new EffectParameter<RGB>("Intermediate Color", dst1);
        this.params.dst2 = new EffectParameter<RGB>("End Color", dst2);
        this.params.duration = new EffectParameter<number>("Duration", duration);
    }
}

export default class TripleCubicColorTransition extends PartialEffect<TripleCubicColorTransitionParams> {
    constructor(params: TripleCubicColorTransitionParams, dt = 0) {
        super(params, dt);
    }
    
    getEffect(t: number) {
        const dt = t - this.params.params.startTime.val;
        let tNorm = 1 - dt / (this.params.params.duration.val);
        tNorm = (tNorm * tNorm * tNorm);

        const dst = tNorm > 0.5 ? this.params.params.dst1 : this.params.params.dst2;
        const src = tNorm > 0.5 ? this.params.params.src : this.params.params.dst1;
        const tNormAdjusted = tNorm <= 0.5 ? (tNorm * 2) : ((tNorm - 0.5) * 2);
        
        return ResolvedEffect.createRgb(ColorTransition.linear(1 - tNormAdjusted, src.val, dst.val));
    }

    isTemporal() {
        return false;
    }

    isComplete(t: number) {
        return false;
    }
}